System/Random.hs

```diff
@@ -240,62 +212,53 @@ mutableByteArrayContentsCompat :: MutableByteArray s -> Ptr Word8
 -- >>> import Control.Monad (replicateM)
 -- >>> import Data.Bits
 -- >>> import Data.Word
--- >>> import System.IO (IOMode(WriteMode), hPutStr, withBinaryFile)
 -- >>> :set -XFlexibleContexts
 -- >>> :set -fno-warn-missing-methods
 
 -- | The class 'RandomGen' provides a common interface to random number
 -- generators.
-{-# DEPRECATED next "Use genWord32[R] or genWord64[R]" #-}
-{-# DEPRECATED genRange "Use genWord32[R] or genWord64[R]" #-}
+{-# DEPRECATED next "No longer used" #-}
+{-# DEPRECATED genRange "No longer used" #-}
 class RandomGen g where
-  {-# MINIMAL (next,genRange)|((genWord32|genWord32R),(genWord64|genWord64R)) #-}
+  {-# MINIMAL split,(genWord32|genWord64|(next,genRange)) #-}
   -- |The 'next' operation returns an 'Int' that is uniformly
   -- distributed in the range returned by 'genRange' (including both
   -- end points), and a new generator. Using 'next' is inefficient as
   -- all operations go via 'Integer'. See
   -- [here](https://alexey.kuleshevi.ch/blog/2019/12/21/random-benchmarks)
-  -- for more details. It is thus deprecated. If you need random
-  -- values from other types you will need to construct a suitable
-  -- conversion function.
+  -- for more details. It is thus deprecated.
   next :: g -> (Int, g)
-  next g = (minR + fromIntegral w, g') where
-    (minR, maxR) = genRange g
-    range = fromIntegral $ maxR - minR
-#if WORD_SIZE_IN_BITS == 32
-    (w, g') = genWord32R range g
-#elif WORD_SIZE_IN_BITS == 64
-    (w, g') = genWord64R range g
-#else
--- https://hackage.haskell.org/package/ghc-prim-0.5.3/docs/GHC-Prim.html#g:1
--- GHC always implements Int using the primitive type Int#, whose size equals
--- the MachDeps.h constant WORD_SIZE_IN_BITS. [...] Currently GHC itself has
--- only 32-bit and 64-bit variants [...].
-# error unsupported WORD_SIZE_IN_BITS
-#endif
+  next g = runGenState g (uniformR (genRange g))
 
   genWord8 :: g -> (Word8, g)
-  genWord8 = first fromIntegral . genWord32R (fromIntegral (maxBound :: Word8))
+  genWord8 = first fromIntegral . genWord32
 
   genWord16 :: g -> (Word16, g)
-  genWord16 = first fromIntegral . genWord32R (fromIntegral (maxBound :: Word16))
+  genWord16 = first fromIntegral . genWord32
 
   genWord32 :: g -> (Word32, g)
-  genWord32 = genWord32R maxBound
+  genWord32 = randomIvalIntegral (minBound, maxBound)
+  -- Once `next` is removed, this implementation should be used instead:
+  -- first fromIntegral . genWord64
 
   genWord64 :: g -> (Word64, g)
-  genWord64 = genWord64R maxBound
+  genWord64 g =
+    case genWord32 g of
+      (l32, g') ->
+        case genWord32 g' of
+          (h32, g'') ->
+            ((fromIntegral h32 `unsafeShiftL` 32) .|. fromIntegral l32, g'')
```

Comment:
This approach assumes that older instances of `RandomGen` class generate 32bits or less of randomness at one time. It will still work for 64bits too, but it will reduce the quality of RNG to 32bits.

Another approach would be to switch to this:

```haskell
genWord64 = randomIvalIntegral (minBound, maxBound)
```
But has terrible performance and is not good default implementation for RNG that define `genWord32`.

Reply:
But this is the default so users can always replace it with something better if it is available?

Reply:
@idontgetoutmuch That was my thinking as well.